### ISSUE TITLE
Timeouts while browsing during first-time scan

### ISSUE THREAD

**Issue by SubJunk** [UniversalMediaServer/UniversalMediaServer#2874]
I noticed yesterday that I was getting timeouts if I browsed during a first-time startup scan, and I think that is something that was better when we had reentrant locks. I'll do some testing before and after to see if it makes a difference.

Any other ideas about how to give priority to the browse actions during scan are welcome. It's a tricky one since the scan basically just "browses" all the shared folders.

**Comment by ik666** [comment]
Didn't we remove all locks? Introducing locks again doesn't sound like a speed update.

If you'd like to change Thread priority, you can do that by `Thread.setPriority()` method, giving the "LibraryScan" a `MIN_PRIORITY` and/or given the RequestV2 Thread a `MAX_PRIORITY `

**Comment by Nadahar** [comment]
I very much doubt thread priority will make much of a difference, it only affects CPU scheduling and this isn't really CPU bound - it's I/O bound. I'm not aware of any mechanism to "prioritize" I/O, I think it's first come first serve for the most part. But, I think I/O will usually be slower the more tasks you do in parallel, as it will spend more time "switching" between the different tasks. In addition it seems like it tends to divide available bandwidth between tasks. So, I think the best way to "prioritize" I/O is to make sure that less important tasks doesn't run when more important ones do. 

HDD's especially can really slow to a crawl if you ask them to perform many transfers in parallel, it's a reason why most file transfers happen sequentially (as opposed to transferring all files in parallel). The drives end up spending a lot of time just moving the read heads back and forth. For this reason it's easy to imagine situations where actually forcing I/O bound tasks to be serialized can actually be faster.

**Comment by SubJunk** [comment]
Ah yeah, I think it's related to Java serialization since I have a lot of log entries about mismatched hashes, which stops caching from happening. I must not have updated or cleared a cache before building, I've had these nightmares before. Or sometimes an IDE does its own serialization and comes up with a new UUID for the same code or something... I'll debug it soon.

**Comment by SubJunk** [comment]
@Nadahar do you remember why this line was added instead of adding a UUID to the file? https://github.com/UniversalMediaServer/UniversalMediaServer/blob/master/src/main/java/net/pms/image/ExifInfo.java#L19

**Comment by Nadahar** [comment]
@SubJunk I assume I was the one who did it..?

I can't say exactly why I did this, but I can speculate as to what might have been my idea. The thing is that the whole serial version ID and Java serialization is a mess.

If you don't specify a serial version ID, Java will generate one from the class using some algorithm. This was originally meant to be "automatic". But, small differences in the JVMs for different platforms (and maybe different versions for all I know) means that the generated serial version ID isn't always the same from one system to another. When people started using serialization for communication over the network or via shared databases, it caused a lot of issues - which is why they introduced the "best practice" that it should be specified manually. 

The problem is that, is that developers have to remember to bump the serial version ID each time changes are made to the class. It's easy to see that not happening, and it would mean that different "versions" of the same class would be considered "compatible" and be deserialized when it really doesn't "fit" the structure of that version of the class. You most likely will still end up with an exception, but a slightly different one. Or, you could risk that it seemingly "worked" but that you get some sneaky invalid instance that's very hard to debug/figure out.

Since these were to be stored in the non-shared database that is unique to each installation, I probably came to the conclusion that it would be better to use the "automatic" serial version ID than to rely on us remembering to bump it manually.

I can't really say what the best thing to do is - ideally some other form of serialization might be preferable. You still don't get around the problem that changes to the class will invalidate stored serialized instances and they must then be rejected and regenerated somehow.

Normally I wouldn't expect that to happen as long as you haven't changed the class, but maybe some changed to the Java version means that the algorithm is changed and that the automatic serial version ID differs. I'm not sure of all the details of the algorithm, but I'm pretty sure this is my logic behind the decision. If you find that using manual serial version IDs causes less "invalidation" of serialized instances, there should be no other downsides from switching to manual IDs than said manual update if the classes are modified.

**Comment by SubJunk** [comment]
That makes sense and matches my understanding of serialization. It also makes sense that this is happening while I'm switching between Java 8 and Java 17 during development - it will be generating different versions of the same files.

The main problem I have now is that it breaks caching and it never recovers; if I restart UMS, the same errors happen for the same files, and so browsing is permanently slow.

```
DEBUG 2022-03-11 20:31:34.285 [Library Scanner] Cached information about video.mkv seems to be from a previous version, reparsing information
TRACE 2022-03-11 20:31:34.285 [Library Scanner] 
java.io.InvalidClassException: net.pms.image.ExifInfo; local class incompatible: stream classdesc serialVersionUID = 7572357352308556233, local class serialVersionUID = 3552335722258681536
	at java.base/java.io.ObjectStreamClass.initNonProxy(Unknown Source)
	at java.base/java.io.ObjectInputStream.readNonProxyDesc(Unknown Source)
	at java.base/java.io.ObjectInputStream.readClassDesc(Unknown Source)
	at java.base/java.io.ObjectInputStream.readNonProxyDesc(Unknown Source)
	at java.base/java.io.ObjectInputStream.readClassDesc(Unknown Source)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject0(Unknown Source)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(Unknown Source)
	at java.base/java.io.ObjectInputStream.readSerialData(Unknown Source)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject0(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject(Unknown Source)
	at org.h2.util.JdbcUtils.deserialize(JdbcUtils.java:430)
	at org.h2.value.ValueToObjectConverter.valueToDefaultObject(ValueToObjectConverter.java:564)
	at org.h2.jdbc.JdbcResultSet.getObject(JdbcResultSet.java:562)
	at net.pms.database.MediaTableFiles.getData(MediaTableFiles.java:516)
	at net.pms.dlna.RealFile.resolve(RealFile.java:230)
	at net.pms.dlna.DLNAResource.syncResolve(DLNAResource.java:1516)
	at net.pms.dlna.DLNAResource.run(DLNAResource.java:1315)
	at net.pms.dlna.RealFile.isValid(RealFile.java:95)
	at net.pms.dlna.DLNAResource.addChild(DLNAResource.java:552)
	at net.pms.dlna.MapFile.manageFile(MapFile.java:290)
	at net.pms.dlna.MapFile.analyzeChildren(MapFile.java:371)
	at net.pms.dlna.RootFolder.scan(RootFolder.java:295)
	at net.pms.dlna.RootFolder.scan(RootFolder.java:256)
	at net.pms.dlna.LibraryScanner.lambda$scanLibrary$0(LibraryScanner.java:41)
	at java.base/java.lang.Thread.run(Unknown Source)
```

I'm looking into it still but happy for any ideas from anyone

**Comment by SubJunk** [comment]
I thought I had fixed it but I'm seeing inconsistency - there seem to be two values that JRE 8 is coming up with for ExifInfo: `3552335722258681536` and `7572357352308556233`. But when I set either of those values, it complains. I either get:

`java.io.InvalidClassException: net.pms.image.ExifInfo; local class incompatible: stream classdesc serialVersionUID = 7572357352308556233, local class serialVersionUID = 3552335722258681536`
or
`java.io.InvalidClassException: net.pms.image.ExifInfo; local class incompatible: stream classdesc serialVersionUID = 3552335722258681536, local class serialVersionUID = 7572357352308556233`

The way I'm testing is to generate a fresh database with an earlier version like 10.14.0, then run [this branch](https://github.com/UniversalMediaServer/UniversalMediaServer/pull/2879)

I'll do more testing and debugging but any ideas are welcome

**Comment by Nadahar** [comment]
As far as I understand it, generating these "default" IDs is really quite pointless. All it does is to generate the value that the Java version that you're running in your IDE at that point would generate "automatically". It doesn't mean that this UID would match what had been automatically generated on other platforms/Java versions.

When I specify manual IDs I simply set it to `1`. After then it must be bumped whenever the class is changed. 

When it comes to `ExifInfo` all implementations are subclasses, since it's an abstract class. So, in reality it's probably the IDs generated for `JPEGinfo`, `RawInfo` and `TIFFInfo` that actually matters. This can explain why you see different IDs - none of them are actually `ExifInfo` instances.

I'm not sure what your goal is. As I see it, the "main problem" is if UMS can't recover from this situation. When these mismatch, they have to be reparsed and the new info should replace the "incompatible" instances. If you have fixed this, the main problem should have been taken care of, and it should work for the users. If the goal is to be able to switch back and forth between different Java versions without the cached instances being invalidated, I'm not sure if that is possible. Even if you specify a fixed ID for all the subclasses, deserialization will fail if there are underlying changes in the "object structure" caused by the different Java version. As I don't know how they generate these, it's impossible to know why different Java versions produces different IDs, but I would assume that it is because something in the underlying structure has changed. If that is the case, there's no getting around having to reparse them, the stored instances simply aren't valid in that Java version.

**Comment by SubJunk** [comment]
@Nadahar given that we almost never (maybe even never in the future) make changes to these files, maybe serialization is overkill. As you can tell I'm inexperienced with serialization, can we easily remove it?

**Comment by Nadahar** [comment]
@SubJunk The reason for using serialization isn't so that they can be easily modified, it's to not make storing them in the database a huge undertaking. To store this information in the database without serializing them, you'd need to create database fields for each and every field in the corresponding classes. Then you'd have to write code that stores each of them and that retrieves them, much like is done for `DLNAMediaInfo` and `DLNAMediaAudio`. In addition you'd have to figure out how to handle the relation between the database entries and the subclasses (which one to create).

Storing them as binary blobs was the only thing that felt "overcomeable" to me. If the information therein would be needed for selects/filtering, binary blobs wouldn't do (at least not for the field(s) in question), but since that's not the case with this information I found it to be the best solution.

You could rewrite it to store each field separately, but it's no small feat.

There are other "non-native" ways to serialize objects. You could serialize them to JSON or XML for example, and use one of the existing libraries to automate this process instead. I don't know how it would compare in speed or storage space, but it should at least avoid the whole ID issue. It could introduce other "challenges" though, in my experience it's hard to predict until you actually try.

**Comment by SubJunk** [comment]
@Nadahar I see, thanks for explaining. I think I understand the motivation behind the serialization now.

I'll work on reproduction steps for the problem I was seeing with the UUID not persisting correctly to the database when changed, and when I'm sure that's either not happening or is fixed, I'll look to give each subclass of ExifInfo its own UUID. I'll try to match the existing ones I have on Windows since that will cover the most users possible, and users with different UUIDs will just need to rescan.

**Comment by SubJunk** [comment]
I think this was partially fixed in https://github.com/UniversalMediaServer/UniversalMediaServer/pull/2896 too for the rootfolder browsing

**Comment by SubJunk** [comment]
It looks like the part that doesn't recover is the TV series thumbnails/posters, so I'm fixing that now

**Comment by ExSport** [comment]
Hi guys.
Long time I wasn't checking UMS progress and see it is still in developmenet, great 👍 
Hopefully all you are well.
For parents I installed new UMS version to share new photos on TV and I saw same issue like years ago and another new one as well :)
I shared one folder with 6000 photos and video files from mobile camera and shared it.
When tried to browse such folder (still not fully scanned=thumbs creation enabled), it timeouts and shows errors, no files (PS3). If restarted UMS during the unfinished scan and restarted it so some analysed files were already in DB, PS3 will not freeze and opens the folder. It seems no (partial) list is sent to PS3 during initial scan but when already some data are in DB, it send it so it will not freeze totally?
Other new issue is persistency. I shared photo folder on external disk and fully scanned it. If you start UMS without USB disk mounted, it seems it resets DB internally as when disk is reattached and UMS started again, it re-analayzes all the files again as hash is not found in DB:
`
18:59:33 TRACE Searching for thumbnail in THUMBNAILS with "SELECT ID FROM THUMBNAILS WHERE MD5 = '77854f146212506e883499c30099c889' LIMIT 1" before update
18:59:33 TRACE Thumbnail "77854f146212506e883499c30099c889" not found in THUMBNAILS
`
Btw. in some issue/pull I saw that UMS in some config updates metadata of files itself. Not talking that this approach can damage file itself but in some cases I share medias from READ ONLY devices/CIFS shares, so such approach can generate exceptions as it expects RW permissions? Didn't check implementation so only mentioning some shares may be ReadOonly.
Another issue is theoretical renderer bandwidth. As I have a lot of "renderers-like" devices (not real renderes like Chrome, Fing,...), UMS will start pinging all of them, at the same time (TV, PS3, ShieldTV,...). On WIFI pinging many devices at one time generates underestimated speeds like 1Mbit for one device and e.g. 70 Mbits for other one, also in case device is on 5Ghz with real speeds of 500 Mbits. I know, test network speed via ICMP packets is not exact at all but something than nothing. But when measuring is done for all renderers at one time, it is totally useless. Nobody will stream from all detected devices at a time but from one or two only... Maybe better to limit speed tests per one renderer at time and test only relevant rederers if all are tested? Btw. it will be nice to see speed under renderer icon like IP, if measured.
Thank you 👍 
Keep up great work guys!
Edit: @valib I wish you good health 👍

**Comment by SubJunk** [comment]
Good to see you @ExSport !
Good thoughts too.

> When tried to browse such folder (still not fully scanned=thumbs creation enabled), it timeouts and shows errors, no files (PS3). If restarted UMS during the unfinished scan and restarted it so some analysed files were already in DB, PS3 will not freeze and opens the folder. It seems no (partial) list is sent to PS3 during initial scan but when already some data are in DB, it send it so it will not freeze totally?

I think this got worse in the last few released versions, I am working hard to get these things resolved. It might be worth trying 10.14.1 since we introduced a few bugs in 10.15.0 that are still not fixed.

> Other new issue is persistency. I shared photo folder on external disk and fully scanned it. If you start UMS without USB disk mounted, it seems it resets DB internally as when disk is reattached and UMS started again, it re-analayzes all the files again as hash is not found in DB:
18:59:33 TRACE Searching for thumbnail in THUMBNAILS with "SELECT ID FROM THUMBNAILS WHERE MD5 = '77854f146212506e883499c30099c889' LIMIT 1" before update 18:59:33 TRACE Thumbnail "77854f146212506e883499c30099c889" not found in THUMBNAILS

This one is tricky because it's a feature that is preventing the database from getting bloated, which would make it slower and take more disk space. Maybe there is a way we can detect removable disks and treat them differently, or have a setting for that feature so it can be disabled.

> Btw. in some issue/pull I saw that UMS in some config updates metadata of files itself. Not talking that this approach can damage file itself but in some cases I share medias from READ ONLY devices/CIFS shares, so such approach can generate exceptions as it expects RW permissions? Didn't check implementation so only mentioning some shares may be ReadOonly.

This might be the upcoming release which can optionally write ID3 tags in music files based on MusicBrainz data. That is disabled by default.

> Another issue is theoretical renderer bandwidth. As I have a lot of "renderers-like" devices (not real renderes like Chrome, Fing,...), UMS will start pinging all of them, at the same time (TV, PS3, ShieldTV,...). On WIFI pinging many devices at one time generates underestimated speeds like 1Mbit for one device and e.g. 70 Mbits for other one, also in case device is on 5Ghz with real speeds of 500 Mbits. I know, test network speed via ICMP packets is not exact at all but something than nothing. But when measuring is done for all renderers at one time, it is totally useless. Nobody will stream from all detected devices at a time but from one or two only... Maybe better to limit speed tests per one renderer at time and test only relevant rederers if all are tested? Btw. it will be nice to see speed under renderer icon like IP, if measured.

Yeah I think these are good points. It's a hard thing to get right, but we can work on it more in these kind of ways.

**Comment by SubJunk** [comment]
The deserialization problem is fixed in https://github.com/UniversalMediaServer/UniversalMediaServer/pull/2905
I have been testing it and it works for me - I am able to change the UUID of `ExifInfo` and `JpegInfo` and it recovers from the error.

I have also merged https://github.com/UniversalMediaServer/UniversalMediaServer/pull/2904 and found it to make a big difference too - if I delete my database and start UMS, and start browsing during the scan, it loads within 1 second, while before it could take 30+ seconds.

So I think this issue can be closed after https://github.com/UniversalMediaServer/UniversalMediaServer/pull/2905 merges

**Comment by SubJunk** [comment]
This is resolved now, details above
